rero_ils/modules/users/api.py

```diff
@@ -0,0 +1,148 @@
+from invenio_userprofiles.models import UserProfile
+from jsonschema import FormatChecker
+from werkzeug.local import LocalProxy
+
+from ...utils import remove_empties_from_dict
+
+_records_state = LocalProxy(lambda: current_app.extensions['invenio-records'])
+
+
+class User(object):
+    """User API."""
+
+    profile_fields = [
+        'first_name', 'last_name', 'street', 'postal_code',
+        'city', 'birth_date', 'username', 'phone', 'keep_history'
+    ]
+
+    def __init__(self, user):
+        """User class initializer."""
+        self.user = user
+
+    @classmethod
+    def create(cls, data, **kwargs):
+        """User record creation.
+
+        :param cls - class object
+        :param data - dictionary representing a user record
+        """
+        with db.session.begin_nested():
+            email = data.pop('email', None)
+            cls._validate(data=data)
+            user = BaseUser(
+                password=hash_password(data.get('birth_date', '123456')),
+                profile=data, active=True)
+            db.session.add(user)
+            profile = user.profile
+            for field in cls.profile_fields:
+                value = data.get(field)
+                if value is not None:
+                    if field == 'birth_date':
+                        value = datetime.strptime(value, '%Y-%m-%d')
+                    setattr(profile, field, value)
+            # send the reset password notification for new users
+            if email:
+                user.email = email
+            db.session.merge(user)
+        db.session.commit()
+        if user.email:
+            send_reset_password_instructions(user)
+        confirm_user(user)
+        return cls(user)
+
+    def update(self, data):
+        """User record update.
+
+        :param data - dictionary representing a user record to update
+        """
+        self._validate(data=data)
+        email = data.pop('email', None)
+        user = self._get_user_by_user_id(self.user.id).user
+        with db.session.begin_nested():
+            if user.profile is None:
+                user.profile = UserProfile(user_id=user.id)
+            profile = user.profile
+            for field in self.profile_fields:
+                if field == 'birth_date':
+                    setattr(
+                        profile, field,
+                        datetime.strptime(data.get(field), '%Y-%m-%d'))
+                else:
+                    setattr(profile, field, data.get(field, ''))
+            # send reset password if email is changed
+            if email and email != user.email:
+                user.email = email
+                send_reset_password_instructions(user)
+            db.session.merge(user)
+        db.session.commit()
+        return self
+
+    @classmethod
+    def _validate(cls, data, **kwargs):
+        """Validate user record against schema."""
+        format_checker = FormatChecker()
+        schema = data.pop('$schema', None)
```

Comment:
Does the validation function correctly? In the schema file the `$schema` is a required field?